### PR TITLE
RGRIDT-687: Implement MarkChangesAsSaved behavior with invalid cells

### PR DESCRIPTION
**Pull request by JoaoFerreira-FrontEnd**
### What was happening
* When the user was calling the MarkChangesAsSaved, we were cleaning the DirtyMarks and the ValidationMarks.

### What was done
* I've changed the API and all the methods including the provider code, so we can choose if we should clean the validation marks or not, using a dynamic flag (boolean).

### Test Steps
1. Build a grid with a SetValidationStatus on the OnCellValueChange;
2. Change the validation status, IsValid value to False;
3. On another button use a new action that uses MarkChangesAsSaved client action;
4. Config the value of ClearInvalids to False;
5. Call this action after changing the value on the previous column;
6. The invalid status mark should continue on the cell;
7. Config the value of ClearInvalids to True;
8. Repeat the test;
9. The invalid status mark shouldn't be displayed;

### Checklist
* [x] tested locally
* [x] documented the code
* [x] clean all warnings and errors of eslint
* [x] requires changes in OutSystems (if so, provide a module with changes)
* [x] requires new sample page in OutSystems (if so, provide a module with changes)

### CODE REVIEW

**Review by tmbp95 — APPROVED** [review]
Nice one!

**Review by glundgren93 — DISMISSED** [review]

File: code/src/GridAPI/GridManager.ts [hunk 1]
```diff
@@ -145,13 +145,17 @@ namespace GridAPI.GridManager {
      *
      * @export
      * @param {string} gridID ID of the Grid where the change will occur.
+     * @param {boolean} clearInvalids determines whether or not we should clean the validation marks.
      */
-    export function MarkChangesAsSaved(gridID: string): void {
+    export function MarkChangesAsSaved(
+        gridID: string,
+        clearInvalids: boolean
```
Comment (rugoncalves) [inline code/src/GridAPI/GridManager.ts:152]: This should have a default value...

**Review by rugoncalves — CHANGES_REQUESTED** [review]

**Review by rugoncalves — APPROVED** [review]
Looking good!

**Review by tmbp95 — APPROVED** [review]

File: code/src/OSFramework/Feature/IValidationMark.ts [hunk 2]
```diff
@@ -5,7 +5,8 @@ namespace OSFramework.Feature {
         invalidRows: Array<any>;
         clear(): void;
         errorMessage(rowNumber: number, binding: string): string;
-        isInvalid(rowNumber: number, binding: string): boolean;
+        isInvalidByRow(row: any): boolean;
+        isInvalidByRowNumber(rowNumber: number, binding: string): boolean;
```
Comment (rugoncalves) [inline code/src/OSFramework/Feature/IValidationMark.ts:9]: I would suggest to keep it as it was. Add only the **isInvalidRow** method.

File: code/src/OSFramework/Grid/IGrid.ts [hunk 3]
```diff
@@ -29,7 +29,7 @@ namespace OSFramework.Grid {
         ): void;
         // eslint-disable-next-line @typescript-eslint/no-explicit-any
         changeProperty(propertyName: string, propertyValue: any): void;
-        clearAllChanges(): void;
+        clearAllChanges(clearInvalids: boolean): void;
```
Comment (rugoncalves) [inline code/src/OSFramework/Grid/IGrid.ts:32]: I think a better name could be, _forceClearDirtyMarks_ what do you think?

File: code/src/WijmoProvider/Features/DirtyMark.ts [hunk 4]
```diff
@@ -164,22 +165,22 @@ namespace WijmoProvider.Feature {
             this._grid.provider.invalidate(); //Mark to be refreshed
         }
 
-        // public clearByRow(row: number): void {
-        //     this._metadata.clearPropertyByRow(row, this._internalLabel);
-        //     this._grid.grid.invalidate(); //Mark to be refreshed
-        // }
+        public clearByRow(row: any): void {
+            this._metadata.clearPropertyByRow(row, this._internalLabel);
```
Comment (rugoncalves) [inline code/src/WijmoProvider/Features/DirtyMark.ts:169]: clearPropertyInRow would be more accurate...

**Review by rugoncalves — COMMENTED** [review]
Other than this good!
I'm thinking since this branch has soo many commits if wouldn't be better to cherry pick the commits or to create a new branch and implement simply the differences.

File: code/src/WijmoProvider/Grid/RowMetadata.ts [hunk 5]
```diff
@@ -56,9 +56,19 @@ namespace WijmoProvider.Grid {
             });
         }
 
-        public clearPropertyByRow(row: number, propertyName: string): void {
-            this.hasOwnProperty(row, propertyName) &&
-                this._getRowMetadata(row).delete(propertyName);
+        public clearPropertyByRow(dataItem: any, propertyName: string): void {
+            // eslint-disable-next-line prettier/prettier
+            if(dataItem.__osRowMetada && dataItem.__osRowMetada.has(propertyName)){
+                dataItem.__osRowMetada.delete(propertyName);
```
[the remark this answers is not in the document]
Comment (JoaoFerreira-FrontEnd) [reply code/src/WijmoProvider/Grid/RowMetadata.ts]: Got it!

File: code/src/OSFramework/Feature/IValidationMark.ts (hunk 2 above)
Comment (JoaoFerreira-FrontEnd) [reply code/src/OSFramework/Feature/IValidationMark.ts:9]: Here, should be like this **isInvalid(rowNumber: number, binding: string): boolean** (original method) and add **isInvalidRow(row: any): boolean** (new need)?

File: code/src/OSFramework/Grid/IGrid.ts (hunk 3 above)
Comment (JoaoFerreira-FrontEnd) [reply code/src/OSFramework/Grid/IGrid.ts:32]: We refer these as ValidationMarks but they are at the same time DirtyMarks.
I think changing the name to  forceClearValidationMarks, probably will be better.

**Pull request closed without merging**